Feature/authentication

serpient opened the pull request:
# Things Done: Add Login Authentication flow to website

### Authentication Strategy
1. User clicks a button which navigates to a server route (/auth/google)
2. Server route does the info transfer with google
3. Server then redirects back to the client login route with a token stored in the cookies (to verify the user against the server for future requests)  and the user UUID stored in the route. 
4. Frontend then parses the information and makes another call to the API for more user information. 
5. User data is then stored in local storage.

### Authenticated / Unauthenticated App + Header
Frontend Authentication strategy uses 2 Apps: AuthenticatedApp (public + private routes) and UnauthenticatedApp (public routes only). Depending on whether there is a "token" in local storage, it will render the correct app which allows access to public vs private routes accordingly.

- Update AuthenticatedApp to render children props, which is used by UnautheticatedApp to add "secure" additional Routes.
- Live navigating and testing in the browser works, and created a unit test that uses MemoryRouter to mock navigate to a /test route and see the rendered page.
- Header component checks localstorage for a user key, and if exists, renders a AuthenticatedHeader which currently has a placeholder avatar and logout route. If user does not exist, then it will render a UnauthenticatedHeader which holds a login link.

serpient committed:
Add login page and google oauth button

serpient committed:
Create Authenticated and Unauthenticated App

Frontend Authentication strategy uses 2 Apps: AuthenticatedApp (public + private routes) and UnauthenticatedApp (public routes only). Depending on whether there is a "token" in local storage, it will render the correct app which allows access to public vs private routes accordingly.

- Update AuthenticatedApp to render children props, which is used by UnautheticatedApp to add "secure" additional Routes.
- Live navigating and testing in the browser works, and created a unit test that uses MemoryRouter to mock navigate to a /test route and see the rendered page.

serpient committed:
Create Authenticated and Unauthenticated Header

Header component checks localstorage for a user key, and if exists, renders a AuthenticatedHeader which currently has a placeholder avatar and logout route. If user does not exist, then it will render a UnauthenticatedHeader which holds a login link.

serpient committed:
Move APIHandler fns into folder

serpient committed:
Update Login Authentication Strategy

Previous Strategy: Allow frontend to communicate with Google, retrieve profile information and token, then return to backend.

Current Strategy: User clicks a button which navigates to a server route (/auth/google), does the info transfer with google, then redirects back to the client login route with a token stored in the cookies (to verify the user against the server for future requests)  and the user UUID stored in the route. From here, the frontend then parses the information and makes another call to the API for more user information. User data is then stored in local storage.

jayden commented on src/components/login/Login.js:
```diff
@@ -0,0 +1,47 @@
+                {handleError(error)}
+                <section className='login--container'>
+                    <h1>Log In</h1>
+                    <a className='google-btn-link' href="http://localhost:4000/auth/google?scope=email%20profile">
```
Since this app is expected to be deployed to Heroku, should this be hardcoded?

serpient replied:
It shouldn't! This was a quick thing to make sure everything was working, but I'll update to switch between localhost and the deployed link.

jayden commented on src/utilities/APIHandler/apiHandler.js:
```diff
@@ -0,0 +1,21 @@
+        this.fetcher = fetcher;
+    }
+
+    getFetcher = () => {
```
Seems strange that this wrapper class exposes a function to get the object that it's wrapping. Do we absolutely need this?

serpient replied:
No it's not necessary -- it was only used in tests to verify it was setting the fetcher correctly. I will remove it

jayden commented on src/utilities/APIHandler/fetcher.js:
```diff
@@ -0,0 +1,26 @@
+        return axios.get(endpoint)
+            .then(response => response.data)
+            .catch(error => {
+                // console.log(error)
```
Can we get rid of these commented out console.log statements?

jayden commented on src/components/header/Header.js:
```diff
@@ -0,0 +1,46 @@
+)
+
+const AuthenticatedHeader = ({ user, setUser }) => {
+    const logOut = (e) => {
```
We can get rid of this parameter since it isn't being used in this function block. Additionally, we can update line 26 as `onClick={logOut}`

jayden commented on src/components/header/Header.js:
```diff
@@ -0,0 +1,46 @@
+        setUser(null);
+    }
+
+    const userAvatar = user.avatar ? user.avatar : "https://ggrmlawfirm.com/wp-content/uploads/avatar-placeholder.png";
```
What would happen if this third party site removes this image from their server? Let's consider serving our own avatar placeholder.

Also, we can simplify this logic by doing:
```javascript
const userAvatar = user.avatar || avatarPlaceholder;
```

jayden commented on src/components/login/Login.js:
```diff
@@ -0,0 +1,47 @@
+                <section className='login--container'>
+                    <h1>Log In</h1>
+                    <a className='google-btn-link' href="http://localhost:4000/auth/google?scope=email%20profile">
+                        <button className='google-btn'>Log In With google</button>
```
Nit-picky, but can we also capitalize Google?

jayden commented on src/App.js:
```diff
@@ -1,25 +1,35 @@
-      </header>
+      <Header user={JSON.parse(localStorage.getItem("user"))} />
       <Router>
+        <Route path="/login/:uuid" exact render={(props) => <Login apiHandler={apiHandler} {...props} />} />
```
We could take advantage of react-router's [Switch](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/Switch.md) component so that we can exclusively render the first route that matches. In that way, we don't need to remember to add the `exact` prop in every route we add.

serpient replied:
The Router has already been changed to the Switch component in the next branch / PR  which branched off this one

jayden reviewed (commented):
Something to think about is that because local storage is accessible to any JavaScript that is run on the site, any malicious code has access to the user's info and token.

zacholauson commented on src/utilities/APIHandler/apiHandler.test.js:
```diff
@@ -44,6 +44,29 @@ test('APIHandler can handle a mock get call that returns an error', async () =>
+        }
+    };
+
+    const apiHandler = new APIHandler(new MockFetcher(itemData));
```
I should have mentioned this back in the other PR... but I noticed that every time we setup this APIHandler in our tests, we build up the MockFetcher with it, is there a specific reason we are mocking the underlying Fetcher and not the APIHandler?

serpient replied:
I had wanted to keep APIHandler as the sole class that the rest of the application interacts with, which felt important to keep consistent when I do larger integration tests. 
But afterring looking at the kinds of tests I AM writing, it does seem like I could just have 2 Handlers: MockAPIHandler and APIHandler. I'll update the PR to reflect that change

zacholauson commented on src/utilities/APIHandler/mockFetcher.js:
```diff
@@ -7,6 +7,11 @@ class MockFetcher {
     }
+
+    post = async (endpoint, data) => {
+        const resolvedData = this.data.error ? this.data : this.data.data;
```
Where is this `data` name coming from? It's kinda strange to read `this.data.data`

serpient replied:
I believe Response returns everything under a `data`, and then our API wraps the specific returned data in another `data` key.
<img width="374" alt="Screen Shot 2019-07-29 at 11 16 10 AM" src="https://user-images.githubusercontent.com/29721784/62071915-5c48fd00-b1f2-11e9-9998-1f465eb430ef.png">

API:
``` elixir
  def render("index.json", %{items: items}) do
    %{data: render_many(items, ItemView, "item.json")}
  end

  def render("show.json", %{item: item}) do
    %{data: render_one(item, ItemView, "item.json")}
  end
```

serpient committed:
Add APIEndpoint URL switcher

API_URL is switches between the local development and deployed API endpoint, depending on whether "REACT_APP_STAGE" env variable is "dev". "REACT_APP_STAGE" is set in the npm scripts for "start"

serpient committed:
Update APIHandler to remove interface

- Instead of a APIHandler with 2 swappable classes of Fetcher and MockFetcher, revise to have 2 classes: APIHandler and MockAPIHandler.

The original interface design was because I thought it would be better for the rest of the codebase to always interact with one class: APIHandler. As opposed to switching between 2 classes depending on whether it was a development or testing environment. However, after looking at the kind of tests I am currently writing, I am not doing the kind of integration tests that requires the rest of the codebase to only know one class. I am injecting the Fetchers directly into each component as a prop.

serpient committed:
Add avatar placeholder asset
